noxfiles/git_nox.py

```diff
@@ -0,0 +1,166 @@
+"""Contains the nox sessions utilities for git-related tasks"""
+
+import re
+from enum import Enum
+from typing import List, Optional
+
+from packaging.version import Version
+
+import nox
+
+RELEASE_BRANCH_REGEX = r"release-(([0-9]+\.)+[0-9]+)"
+RC_TAG_REGEX = r"{release_version}rc([0-9]+)"
+RELEASE_TAG_REGEX = r"(([0-9]+\.)+[0-9]+)"
+VERSION_TAG_REGEX = "{version}{tag_type}([0-9]+)"
+
+# our tag "types"
+
+
+class TagType(Enum):
+    RC = "rc"  # used for release branches
+    ALPHA = "a"  # used for feature branches
+    BETA = "b"  # used for `main` branch
+
+
+def get_all_tags(repo):
+    """
+    Returns a list of all tags in the repo, sorted by committed date, latest first
+    """
+
+    git_session = repo.git()
+    git_session.fetch("--force", "--tags")
+    return sorted(repo.tags, key=lambda t: t.commit.committed_datetime, reverse=True)
+
+
+def get_current_tag(
+    existing: bool = False, repo=None, all_tags: List = []
+) -> Optional[str]:
+    """
+    Get the current git tag.
+
+    If `existing` is true, this tag must already exist.
+    Otherwise, a tag is generated via `git describe --tags --dirty --always`,
+    which includes "dirty" tags if the working tree has local modifications.
+    """
+    if repo is None:
+        from git.repo import Repo
+
+        repo = Repo()
+    if not all_tags:
+        all_tags = get_all_tags(repo)
+    if existing:  # checks for an existing tag on current commit
+        return next(
+            (tag.name for tag in all_tags if tag.commit == repo.head.commit),
+            None,
+        )
+    git_session = repo.git()
+    git_session.fetch("--force", "--tags")
+    current_tag = git_session.describe("--tags", "--dirty", "--always")
+    return current_tag
+
+
+@nox.session()
+def tag(session: nox.Session) -> str:
+    """
+    Generates and applies a git tag based on the currently checked out branch
+    and existing tags in the repo.
+
+    Positional Arguments:
+    - dry_run = don't actually apply the tag or push, just show the tag that will be generated
+    - push = push the generated tag
+    - push_current = push the current tag
+    """
+    from git.repo import Repo
+
+    repo = Repo()
+    all_tags = get_all_tags(repo)
+    current_tag = get_current_tag(
+        existing=True,
+        repo=repo,
+        all_tags=all_tags,
+    )  # first get an existing tag if we've got one
+    generated_tag = generate_tag(
+        session, repo.active_branch.name, all_tags
+    )  # generate a tag based on the current repo state
+
+    if "dry_run" in session.posargs:
```

Comment:
I really like the `dry_run` 🙌🏽 

I noticed there was some hesitancy about pushing too many tags up etc. Does it make sense to have some sort of guardrails around not using the `dry_run` ? Some extra option like "to continue enter y" sort of thing or to have `dry_run` be the default and a separate positional argument to run it through? Also fine to leave as is, just thinking out loud

Reply:
hmm i may not be totally tracking what you're suggesting, but just to be clear on the behavior here:
- if no argument is passed, i.e. `nox -s tag`, then a tag is applied to the current commit locally but not pushed
- if `nox -s tag -- dry_run` is run (or if `dry_run` is included with any other additional arguments too), then no tag is applied, you just see the tag that _would_ be applied to the current commit
- if `nox -s tag -- push` is run, then a tag is generated and applied locally _and_ pushed up to the remote repo
- if `nox -s tag -- push_current` is run, then no tag is generated or applied locally, but instead we just try to push up whatever the current commit is tagged as locally

i felt like that covered the bases pretty well and defaulted to applying, but not pushing, a tag, which feels like a reasonable default because it gives you _something_, i.e. it generates and applies a tag, but it means you've got to be deliberate about actually _pushing_ that tag, which is the action with more consequences. that being said, i'm definitely open to alternatives, and it should be easy to tweak this logic if we'd like

Reply:
Gotcha, that makes sense - thanks for laying that out! 

I wasn't sure if having a bunch of tags created locally is something we wanted or needed to avoid but it sounds like the `nox -s tag -- push` will generate a tag of the current state regardless and push it. So although there may be a lot of tags generated locally it doesn't matter as much or add any risk around having them pushed to the repo, right?

Reply:
yeah, that's generally true, although if you generate a bunch of tags locally, they'll keep on getting incremented unnecessarily. that's not super consequential, but it's probably less than ideal. maybe i'll switch the default to actually be the dry_run, since that's truly the least intrusive thing, and then there can be a `push` flag that is used to actually publish your changes. the tagging without pushing is not very useful and does have some less-than-ideal side effects if done without care, so i think you're right that it shouldn't be the default.

this is making sense to me, i'll push up a change in a bit and let me know what you think! happy to talk it over if my stream of consciousness above is hard to follow! 😄